src/org/exist/xquery/ExternalModuleImpl.java

```diff
@@ -268,9 +270,25 @@ public boolean moduleIsValid(DBBroker broker) {
         return mSource != null && mSource.isValid(broker) == Source.VALID;
     }
 
-    public void reset(XQueryContext xqueryContext) {
-        mContext.reset();
-        mStaticVariables.clear();
+    @Override
+    public void reset(XQueryContext context) {
+        // deprecated, ignore
+    }
+
+    public void reset(XQueryContext xqueryContext, boolean keepGlobals) {
+        // prevent recursive calls by checking needsReset
+        if (needsReset) {
+            needsReset = false;
+            mContext.reset(keepGlobals);
+            if (!keepGlobals) {
+                mStaticVariables.clear();
+                // reset state of variable declarations
+                mGlobalVariables.values().forEach(v -> v.resetState(true));
+            }
+            // reset state of declared functions
+            mFunctionMap.values().forEach(f -> f.resetState(true));
+            needsReset = true;
```

Comment:
Can you explain why when `needsReset` is `true`, it is set to `false` and then back to `true` again?